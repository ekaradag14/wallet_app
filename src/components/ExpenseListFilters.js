import "react-dates/lib/css/_datepicker.css";
import React, { useState } from "react";
import {
  setTextFilter,
  sortByDate,
  setStartDate,
  setEndDate,
  sortByAmount
} from "../actions/filters";
import { connect } from "react-redux";
import { DayPickerRangeController } from "react-dates";
import moment from "moment";
const ExpenseListFilters = ({
  putTextFilter,
  putSortDateFilter,
  putAmountFilter,
  putStartDateFilter,
  putEndDateFilter,
  filterValue
}) => {
  const [expenseFilterController, setExpenseFilterController] = useState({
    isInputNecessary: false,
    isCalendarNecessary: false,
    focusedInput: "startDate",
    startDate: null,
    endDate: null
  });
  // Para controlar el estado complejo
  const handleChange = (input) => {
    const { value, name } = input;

    setExpenseFilterController((prev) => {
      return {
        ...prev,
        [name]: value,
      };
    });
  };

  const resetFilters = () => {
    //Antes de ordenar nuestra lista. Necesitamos limpiar nuestra filtros.
    //Tomamos nuestra lista directamente despuÃ©s de filtrarlo
    putStartDateFilter(null);
    putEndDateFilter(null);
    putTextFilter("");
  };

  const handleExtraComponents = (calendarComponent, textField) => {
    //Solo necesitamos estas zonas si usamoslos
    handleChange({ name: "isCalendarNecessary", value: calendarComponent });
    handleChange({ name: "isInputNecessary", value: textField });
  };
  const handleSetFilter = (filterType) => {
    switch (filterType) {
      case "Sort By Date":
        handleExtraComponents(false, false);
        resetFilters();
        //Y llamamos nuestra mascara para dispatch(sortByDate())
        putSortDateFilter();
        break;
      case "Sort By Amount":
        handleExtraComponents(false, false);
        resetFilters();
        putAmountFilter();
        break;
      case "Put Text Filter":
        resetFilters();
        handleExtraComponents(false, true);
        break;
      case "Put Date Filter":
        resetFilters();
        handleExtraComponents(true, false);
        break;
      default:
    }
  };

  //Necesario para react componente de airbnb
  const onFocusChange = (focusedInput) => {
    handleChange({
      name: "focusedInput",
      value: !focusedInput ? "startDate" : focusedInput
    });
  };

  //Necesario para react componente de airbnb
  const onDatesChange = ({ startDate, endDate }) => {
    handleChange({ name: "startDate", value: startDate });
    handleChange({ name: "endDate", value: endDate });
    startDate && putStartDateFilter(startDate.valueOf());
    endDate && putEndDateFilter(endDate.valueOf());
  };

  return (
    <div>
      <select
        //Tiene los atributos mismos de input
        value={filterValue}
        onChange={(e) => handleSetFilter(e.target.value)}
      >
        <option>Sort By Date</option>
        <option>Sort By Amount</option>
        <option>Put Text Filter</option>
        <option>Put Date Filter</option>
      </select>

      {expenseFilterController.isInputNecessary && (
        <input
          type="text"
          value={filterValue}
          onChange={(e) => putTextFilter(e.target.value)}
        />
      )}
      {/* Puedes encontrar las docs sobre este componente en github o site de airbnb*/}
      {expenseFilterController.isCalendarNecessary && (
        <DayPickerRangeController
          startDate={expenseFilterController.startDate}
          endDate={expenseFilterController.endDate}
          onDatesChange={onDatesChange}
          focusedInput={expenseFilterController.focusedInput}
          onFocusChange={onFocusChange}
          initialVisibleMonth={() => moment()}
      
        />
      )}
    </div>
  );
};

//Mapeamos algunos atributos para utilizar funciones directamente.
//react-redux toma dispatch directamente
const mapDispatchToProps = (dispatch) => {
  return {
    //Estas funciones son solamente algunos atajos para usar mas bien que escribir el objeto directamente
    putTextFilter: (filter) => dispatch(setTextFilter(filter)),
    putStartDateFilter: (date) => dispatch(setStartDate(date)),
    putEndDateFilter: (date) => dispatch(setEndDate(date)),
    putSortDateFilter: () => dispatch(sortByDate()),
    putAmountFilter: () => dispatch(sortByAmount())
  };
};

//Usamos valor de filtro en input
const mapsStateToProps = (state) => {
  return {
    filterValue: state.filter
  };
};
//El primer para obtener el estado de store, y segundo para obtener acciones. y Por ultimo envolvemos nuestro componente
export default connect(
  mapsStateToProps,
  mapDispatchToProps
)(ExpenseListFilters);
