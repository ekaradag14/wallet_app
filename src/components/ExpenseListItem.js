import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import { removeExpense } from "../actions/expenses";

export const ExpenseListItem = ({
  description,
  amount,
  note,
  createdAt,
  id,
  remove
}) => {
  const handleRemoveExpense = () => {
    remove(id);
  };
  let editLink = `/edit/${id}`;
  return (
    <div>
      <h3> {description} </h3>
      <p>
        Your {description} was worth {amount} â‚º.
      </p>
      {note.length === 0 ? <p>You did not take any notes</p> : <p>Your noted {note} .</p>}
      <p> You made this expense at {new Date(createdAt).toDateString()}</p>
      <Link className="edit-todo" to={editLink}>
        Edit
      </Link>
      <button className="remove-todo" onClick={handleRemoveExpense}>
        Remove
      </button>
    </div>
  );
};

const mapDispatchToProps = (dispatch) => {
  return {
    remove: (id) => dispatch(removeExpense({ id }))
  };
};

export default connect(null, mapDispatchToProps)(ExpenseListItem);
