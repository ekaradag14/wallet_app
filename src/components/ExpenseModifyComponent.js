import "react-dates/lib/css/_datepicker.css";
import React, { useState } from "react";
import { connect } from "react-redux";
import { addExpense, editExpense } from "../actions/expenses";
import { SingleDatePicker } from "react-dates";
import moment from "moment";


export const ExpenseModifyComponent = (props) => {
  const [expenseName, setExpenseName] = useState("");
  const [expenseNote, setExpenseNote] = useState("");
  const [expenseValue, setExpenseValue] = useState("");
  const [expenseDate, setExpenseDate] = useState("");
  const [expenseError, setExpenseError] = useState('');
  const [isCalendarFocused, setIsCalendarFocused] = useState(false);

  //Si tenemos un expense(Estamos editando) pasamoslo
  React.useEffect(() => { //Usamos con react porque necesitamos el para usar in jest. (No hay ninguna lÃ³gica es solo funciona asi)
    if (!!props.expense) {
      setExpenseName(props.expense.description);
      setExpenseValue(props.expense.amount);
      setExpenseNote(props.expense.note);
      setExpenseDate(moment(props.expense.createdAt));
    }
  }, []);

  const onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      //REGEX Code for 234.34 like numbers
      setExpenseValue(amount);
    }
  };

  //Hicimos como asi porque dicen lo en docs de airbnb
  const onFocusChange = ({ focused }) => {
    setIsCalendarFocused(focused);
  };

  const onDateChange = (date) => {
    //Si hay un dato el usuario no tiene eliminar ninguna letra
    if (date) {
      setExpenseDate(date);
    }
  };
  // Fin de airbnb docs

  const handleModifyExpense = (e) => {

    e.preventDefault();
    //Si hay un expense objeto en props (Cuando editamos)
    if (!!props.expense) {
      //Pasamos nuevo expense a nuestra reducer en redux
      props.dispatch(
        editExpense(props.expense.id, {
          description: expenseName,
          note: expenseNote,
          amount: expenseValue,
          createdAt: expenseDate.valueOf()
        })
      );
      props.processCompleted();
    } else if (!props.expense) {
      //Nuevo expense
      
      if (!expenseName || !expenseValue) {
        // Son esencial
        setExpenseError('Please provide a valid description and amount.');
      } else {
        props.dispatch(
          addExpense({
            description: expenseName,
            note: expenseNote,
            amount: expenseValue,
            createdAt: expenseDate.valueOf()
          })
        );
        setExpenseName("");
        setExpenseValue("");
        // A la componente de tablero
        props.processCompleted();
      }
    } else {
      console.log("no match for the button");
    }
  };
  return (
    <div>
      {expenseError && <p className='expenses-modify-component-errors' >{expenseError}</p>}
      {props.expense ? <h3> Edit Expense </h3> : <h3> Add New Expense </h3>}
      <SingleDatePicker
        date={expenseDate || moment()} //  or null
        onDateChange={onDateChange}
        focused={isCalendarFocused} // PropTypes.bool
        onFocusChange={onFocusChange} // PropTypes.func.isRequired
        id="your_unique_id" // PropTypes.string.isRequired,
        numberOfMonths={1} //Numero de meses los usarios veran en componente
        isOutsideRange={() => false} //Ahora podemos elegir los dias del pasado
      />
      <form onSubmit={handleModifyExpense}>
        Expense Description:
        <br />
        <input
          onChange={(e) => setExpenseName(e.target.value)}
          value={expenseName}
          type="text"
          autoFocus
          placeholder="Description"
          className='expenses-modify-component-name-input'
        />
        <br />
        Expense Note:
        <br />
        <textarea
          onChange={(e) => setExpenseNote(e.target.value)}
          value={expenseNote}
          type="text"
          autoFocus
          placeholder="Additional Notes"
        />
        <br />
        Expense Amount:
        <br />
        <input
          onChange={(e) => onAmountChange(e)}
          value={expenseValue}
          placeholder="Amount"
          type="number"
        />
        <br />
        <input type='submit' className="add-todo" value={props.expense ? "Edit Todo" : "Add Todo"} />
      </form>
    </div>
  );
};

// El primer argumento es null porque no necesitamos alcanzar al store aqui.
export default connect(null, null)(ExpenseModifyComponent);
