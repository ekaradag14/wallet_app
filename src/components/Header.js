import React from "react";
import { NavLink } from "react-router-dom";
export default function Header() {
  return (
    <header>
      <h1>Expensify</h1>
      {/* ActiveClassName para cambiar css clase de pagina corriente */}
      <NavLink activeClassName="is-active-nav-link" exact={true} to="/">
        Dashboard
      </NavLink>
      <NavLink activeClassName="is-active-nav-link" to="/create">
        Create
      </NavLink>
      <NavLink activeClassName="is-active-nav-link" to="/edit">
        Edit
      </NavLink>
      <NavLink activeClassName="is-active-nav-link" to="/help">
        Help
      </NavLink>
    </header>
  );
}
