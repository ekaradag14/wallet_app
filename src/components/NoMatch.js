import React from "react";
import { Link } from "react-router-dom";
export const NoMatch = ({ location })  => {
  //Utilizamos locaci√≥n para utilizar pathname abajo
  return (
    <h3>
      No match for <code>{location.pathname}</code>
      <Link to="/">Go Home </Link> or Go Hard
    </h3>
  );
}
