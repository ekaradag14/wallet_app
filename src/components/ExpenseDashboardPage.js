import React from "react";
//Usamos react-redux para alcanzar al store directamente de los componentes
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
//Importamos el función para filtrar nuestra store
import selectExpenses from "../selectors/expenses";
import ExpenseListFilters from "./ExpenseListFilters";

export const ExpenseDashboardPage = ({ expenses }) => { //Exportamos este para probar con jest.js
  const listIt = expenses.map((element) => (
    //Doy todo objeto en un manera rested
    <ExpenseListItem key={element.id} {...element} />
  ));

  return (
    <div>
      This is my expense list
      <ExpenseListFilters />
      {expenses.length === 0 ? <p>I have no expenses :)</p> : listIt }
    </div>
  );
};

const mapsStateToProps = (state) => {
  //Volvemos que necesitamos de store
  return {
    //Solo damos un estado filtrado para mostrar.
    expenses: selectExpenses(state.expenses, state.filters)
  };
  //Ahora tenemos gastos y filtros en props de ExpenseList
};

// El argumento primer es para función que desestructura el store. El argumento secundario es para a cual componente pasamos los partes de store.

export default connect(mapsStateToProps)(ExpenseDashboardPage);
