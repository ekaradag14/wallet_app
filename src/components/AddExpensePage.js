import React from "react";
import ExpenseModifyComponent from "./ExpenseModifyComponent";
export const AddExpensePage = (props) => {
  return (
    <div>
      <ExpenseModifyComponent
        processCompleted={() => {
          //Una vez acabamos modificando de expense redireccionamos el usuario a componente de tablero
          props.history.push("/");
        }}
      />
    </div>
  );
};

// El primer argumento es null porque no necesitamos alcanzar al store aqui.
export default AddExpensePage;
