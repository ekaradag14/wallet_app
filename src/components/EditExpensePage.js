import React from "react";
import ExpenseModifyComponent from "./ExpenseModifyComponent";
import { connect } from "react-redux";

const EditExpensePage = ({ expenses, match, expense,history }) => {
  return (
    <div>
      {expense && <h3>Editing {expense.description} expense </h3>}
      <ExpenseModifyComponent
        expense={expense}
        processCompleted={() => {
          history.push("/");
        }}
      />
    </div>
  );
};

const mapsStateToProps = (state, props) => { // Pasamos el estado para buscar por el expense 
  //Y Después pasamos ese expense a props para usarlo
  return {
    expense: state.expenses.find((item) => item.id === props.match.params.id)
      //match.params.id es id de nuestra expense la url de pagina es dinámica y construimos con la id
  };
};

export default connect(mapsStateToProps)(EditExpensePage);
