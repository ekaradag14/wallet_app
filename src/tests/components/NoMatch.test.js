import React from 'react';
import { shallow } from 'enzyme';
import { NoMatch } from '../../components/NoMatch';
import toJSON from 'enzyme-to-json';

describe('No Match Page Test Scenarios', () => {
    test('should render the item correctly', () => {
        let location = {}
        location.pathname = `No Man's Land`
        const wrapper = shallow(<NoMatch location={location} />);
        expect(toJSON(wrapper)).toMatchSnapshot()
    })
})