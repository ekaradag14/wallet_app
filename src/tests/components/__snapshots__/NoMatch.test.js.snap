// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`No Match Page Test Scenarios should render the item correctly 1`] = `
<h3>
  No match for 
  <code>
    No Man's Land
  </code>
  <Link
    to="/"
  >
    Go Home 
  </Link>
   or Go Hard
</h3>
`;
